File: Tree/113PathSumII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        vector<int> temp;
        helper(root, temp, res, sum);
        return res;
    }
    void helper(TreeNode* root, vector<int> temp, vector<vector<int>>& res, int sum){
        if(!root) return;
        temp.push_back(root->val);
        if(!root->left && !root->right && root->val == sum){
            res.push_back(temp);
            return;
        }
        helper(root->left, temp, res, sum-root->val);
        helper(root->right, temp, res, sum-root->val);
    }
};

File: Tree/113PathSumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(res, [], root, sum)
        return res
    
    def dfs(self, res, tmp, root, sum):
        if not root:
            return
        if not root.left and not root.right and root.val == sum:
            tmp.append(root.val)
            res.append(tmp)
            return
        
        self.dfs(res, tmp + [root.val], root.left, sum - root.val)
        self.dfs(res, tmp + [root.val], root.right, sum - root.val)